### PR TITLE
A ClientBuilderBase for REST-based client wrappers

### PR DESCRIPTION
**Pull request by jskeet**
Specifically, this will be used by Storage, Translation and BigQuery.

Currently no tests - I wanted to get feedback on the shape first. Feedback on what aspects you'd want to see tests on would be useful, too...

### CODE REVIEW

**Comment by jskeet** [comment]
I note that the test failure is BufferedClientStreamWriterTest, which has been more flaky than usual recently. We might want to look into that. Will restart Travis for now...

File: Google.Api.Gax.Rest/ClientBuilderBase.cs [hunk 1]
```diff
@@ -0,0 +1,192 @@
...
+        protected void ValidateOptionExcludesOthers(string message, object controlling, params object[] values)
+        {
+            GaxPreconditions.CheckState(controlling == null || values.All(v => v == null), message);
+        }
+
+        /// <summary>
+        /// Creates an initializer for the service. This method does not perform any validation.
+        /// </summary>
+        /// <returns>An initializer for the service.</returns>
+        protected BaseClientService.Initializer CreateServiceInitializer() =>
```
Comment (chrisdunelm) [inline Google.Api.Gax.Rest/ClientBuilderBase.cs]: `virtual`?

File: Google.Api.Gax.Rest/ClientBuilderBase.cs [hunk 2]
```diff
@@ -0,0 +1,192 @@
...
+            {
+                // Don't use credentials if we're using an API key.
+                return null;
+            }
+            if (Credential != null)
+            {
+                return Credential;
+            }
+            GoogleCredential unscoped =
+                CredentialsPath != null ? GoogleCredential.FromFile(CredentialsPath) : // TODO: Use an async method when one is available
```
Comment (chrisdunelm) [inline Google.Api.Gax.Rest/ClientBuilderBase.cs]: Comment isn't relevant here :)

File: Google.Api.Gax.Rest/ClientBuilderBase.cs [hunk 3]
```diff
@@ -0,0 +1,192 @@
...
+
+        // Note: The implementation is responsible for performing validation before constructing the client.
+        // The Validate method can be used as-is for most builders.
+
+        /// <summary>
+        /// Builds the resulting client.
+        /// </summary>
+        public abstract TClient Build();
+
+        // Note: The implementation is responsible for performing validation before constructing the client.
```
Comment (chrisdunelm) [inline Google.Api.Gax.Rest/ClientBuilderBase.cs]: Duplicate comment?

**Comment by jskeet** [comment]
Note to self: ApiKey + credentials is probably a scenario we *do* want to allow. We should probably just not fetch the default credentials implicitly if there's an API key, but allow any credentials to be specified explicitly.

**Comment by jskeet** [comment]
Okay, rewritten a little to handle API keys differently, and added tests. I believe it's ready to merge now - PTAL when you can.

File: Google.Api.Gax.Grpc/Google.Api.Gax.Grpc.csproj [hunk 4]
```diff
@@ -20,7 +20,7 @@
 
     <PackageReference Include="Grpc.Auth" Version="1.20.1" />
     <PackageReference Include="Grpc.Core" Version="1.20.1" />
-    <PackageReference Include="Google.Apis.Auth" Version="1.39.0" />
+    <PackageReference Include="Google.Apis.Auth" Version="1.40.0" />
```
Comment (chrisdunelm) [inline Google.Api.Gax.Grpc/Google.Api.Gax.Grpc.csproj]: Probably worth updating to 1.40.2

File: Google.Api.Gax.Rest/ClientBuilderBase.cs [hunk 5]
```diff
@@ -0,0 +1,181 @@
...
+
+        /// <summary>
+        /// The credential to use for authentication. This cannot be specified alongside other authentication properties.
+        /// </summary>
+        public ICredential Credential { get; set; }
+
+        /// <summary>
+        /// An API key to use instead of a regular credential. If this is non-null and no other credentials are supplied,
+        /// it will be used as the only credentials. If other credentials are supplied (such as through <see cref="CredentialsPath"/>)
+        /// then the two
```
Comment (chrisdunelm) [inline Google.Api.Gax.Rest/ClientBuilderBase.cs]: Unfinished sentence.

File: Google.Api.Gax.Rest/ClientBuilderBase.cs [hunk 6]
```diff
@@ -0,0 +1,181 @@
...
+            {
+                return Credential;
+            }
+            GoogleCredential unscoped =
+                CredentialsPath != null ? GoogleCredential.FromFile(CredentialsPath) :
+                JsonCredentials != null ? GoogleCredential.FromJson(JsonCredentials) :
+                null; // Use default credentials (maybe - see below)
+            // While we accept any credentials that are specified even if there's an API key,
+            // we don't try to load default credentials.
+            if (ApiKey != null && unscoped is null)
```
Comment (chrisdunelm) [inline Google.Api.Gax.Rest/ClientBuilderBase.cs]: Really need an `isnt` operator for consistency.

File: Google.Api.Gax.Rest/Google.Api.Gax.Rest.csproj [hunk 7]
```diff
@@ -16,7 +17,7 @@
   <ItemGroup>
     <ProjectReference Include="..\Google.Api.Gax\Google.Api.Gax.csproj" />
 
-    <PackageReference Include="Google.Apis.Auth" Version="1.39.0" />
+    <PackageReference Include="Google.Apis.Auth" Version="1.40.0" />
```
Comment (chrisdunelm) [inline Google.Api.Gax.Rest/Google.Api.Gax.Rest.csproj]: 1.40.2

File: Google.Api.Gax.Grpc/Google.Api.Gax.Grpc.csproj (hunk 4 above)
Comment (jskeet) [reply Google.Api.Gax.Grpc/Google.Api.Gax.Grpc.csproj]: Will do.

File: Google.Api.Gax.Rest/ClientBuilderBase.cs (hunk 5 above)
Comment (jskeet) [reply Google.Api.Gax.Rest/ClientBuilderBase.cs]: Whoops, thanks. Will fix.

File: Google.Api.Gax.Rest/ClientBuilderBase.cs (hunk 6 above)
Comment (jskeet) [reply Google.Api.Gax.Rest/ClientBuilderBase.cs]: Ack. `ApiKey is object` works, but it's a bit odd-looking to start with.

File: Google.Api.Gax.Rest/Google.Api.Gax.Rest.csproj (hunk 7 above)
Comment (jskeet) [reply Google.Api.Gax.Rest/Google.Api.Gax.Rest.csproj]: Done.

**Pull request merged**